### PR TITLE
Use Typedoc for Wasm API docs

### PR DESCRIPTION
**Pull request by eike-hass**
# Description of change
Please write a summary of your changes and why you made them.

## Links to any relevant issues
Be sure to reference any related issues by adding `fixes issue #`.

## Type of change
Add an `x` to the boxes that are relevant to your changes.

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [ ] Enhancement (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Fix

## How the change has been tested
Describe the tests that you ran to verify your changes.
Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist
Add an `x` to the boxes that are relevant to your changes.

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by UMR1352 — APPROVED** [review]

File: .github/workflows/shared-build-wasm.yml [hunk 1]
```diff
@@ -89,5 +89,6 @@ jobs:
             bindings/wasm/node
             bindings/wasm/web
             bindings/wasm/examples/dist
```
Comment (Dr-Electron) [inline .github/workflows/shared-build-wasm.yml:91]: This can be removed, no?
```suggestion
```

File: .github/workflows/upload-docs.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  release:
+    types: [published]
+  workflow_dispatch:
+    inputs:
+      version:
+        description: 'Version to publish docs under (e.g. `v1.2.3-dev.1`)'
+        required: true
+
+env:
+  GH_TOKEN: ${{ github.token }}
```
Comment (Dr-Electron) [inline .github/workflows/upload-docs.yml:13]: I can't remember, but I think this token has to be generated per repo? 🤔

File: bindings/wasm/typedoc.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    "githubPages": false,
+    "theme": "markdown",
+    "entryDocument": "api_ref.md",
+    "hideBreadcrumbs": true,
+    "hideGenerator": true,
+    "sort": ["source-order"],
+    "compilerOptions": {
+        "skipLibCheck": true,
+    }
+}
```
Comment (Dr-Electron) [inline bindings/wasm/typedoc.json:23]:
```suggestion
}

```

File: bindings/wasm/package.json [hunk 4]
```diff
@@ -71,6 +71,8 @@
     "ts-node": "^10.9.1",
     "tsconfig-paths": "^4.1.0",
     "txm": "^8.1.0",
+    "typedoc": "^0.24.6",
+    "typedoc-plugin-markdown": "^3.14.0",
```
Comment (Dr-Electron) [inline bindings/wasm/package.json:75]: We should update them some day. But for now I only got it working with these versions 😅

File: .github/workflows/upload-docs.yml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    with:
+      run-unit-tests: false
+      ref: ${{ inputs.ref }}
+      output-artifact-name: identity-docs
+
+  upload-docs:
+    runs-on: ubuntu-latest
+    needs: build-wasm
+    steps:
+      - uses: actions/download-artifact@v2
```
Comment (Dr-Electron) [inline .github/workflows/upload-docs.yml:31]: Should also be updated to v4. As even v3 will be deprecated this year. But we can do that in another PR too

**Review by Dr-Electron — CHANGES_REQUESTED** [review]

File: .github/workflows/shared-build-wasm.yml (hunk 1 above)
Comment (eike-hass) [reply .github/workflows/shared-build-wasm.yml:91]: I don't think so, that artifact is used for the cypress e2e test

File: .github/workflows/upload-docs.yml (hunk 2 above)
Comment (eike-hass) [reply .github/workflows/upload-docs.yml:13]: what is the token used for?

File: .github/workflows/upload-docs.yml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+          fi
+          VERSION=$(echo $INPUT_VERSION | sed -e 's/.*v\([0-9]*\.[0-9]*\).*/\1/')
+          echo VERSION=$VERSION >> $GITHUB_OUTPUT
+      - name: Compress generated docs
+        run: |
+          tar czvf wasm.tar.gz docs/*
+
+      - name: Upload docs to AWS S3
+        env:
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID_IOTA_WIKI }}
```
Comment (eike-hass) [inline .github/workflows/upload-docs.yml:50]: those secrets still need to be created in the repo, right?

File: .github/workflows/upload-docs.yml (hunk 2 above)
Comment (Dr-Electron) [reply .github/workflows/upload-docs.yml:13]: If I remeber correctly you have to specify it for checkout I think. If the trigger was created by another action as then the default github token won't be valid.

So in our case whatever workflow you have to publish a release will run with the "default" token and will trigger this action, which then has to specify the token explicitly

File: .github/workflows/upload-docs.yml (hunk 6 above)
Comment (eike-hass) [reply .github/workflows/upload-docs.yml:50]: It seems to work, so we are all good, right?

File: .github/workflows/upload-docs.yml (hunk 5 above)
Comment (eike-hass) [reply .github/workflows/upload-docs.yml:31]: fine to update it here

File: .github/workflows/upload-docs.yml (hunk 2 above)
Comment (eike-hass) [reply .github/workflows/upload-docs.yml:13]: then all our other shared workflows wouldn't work 🤔 I think this fine to remove

**Review by wulfraem — APPROVED** [review]

File: .github/workflows/upload-docs.yml (hunk 5 above)
Comment (Dr-Electron) [reply .github/workflows/upload-docs.yml]: 4731a6fc1e9dfed981f078295b729d77b7b329a8

File: .github/workflows/upload-docs.yml (hunk 2 above)
Comment (Dr-Electron) [reply .github/workflows/upload-docs.yml]: I removed it. If it doesn't work it is your fault and we have to add it back after the first release :trollface:

File: .github/workflows/build-and-test.yml [hunk 7]
```diff
@@ -170,7 +170,7 @@ jobs:
     needs: check-for-run-condition
     if: ${{ needs.check-for-run-condition.outputs.should-run == 'true' }}
     # owner/repository of workflow has to be static, see https://github.community/t/env-variables-in-uses/17466
-    uses: iotaledger/identity.rs/.github/workflows/shared-build-wasm.yml@main
+    uses: iotaledger/identity.rs/.github/workflows/shared-build-wasm.yml@typedoc
```
Comment (Dr-Electron) [inline .github/workflows/build-and-test.yml:173]:
```suggestion
    uses: iotaledger/identity.rs/.github/workflows/shared-build-wasm.yml@main
```

File: .github/workflows/upload-docs.yml [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        description: 'Version to publish docs under (e.g. `v1.2.3-dev.1`)'
+        required: true
+
+permissions:
+  actions: 'write'
+
+jobs:
+  build-wasm: 
+    # owner/repository of workflow has to be static, see https://github.community/t/env-variables-in-uses/17466
+    uses: iotaledger/identity.rs/.github/workflows/shared-build-wasm.yml@typedoc
```
Comment (Dr-Electron) [inline .github/workflows/upload-docs.yml:18]:
```suggestion
    uses: iotaledger/identity.rs/.github/workflows/shared-build-wasm.yml@main
```

File: .github/workflows/wasm-automatic-release-and-publish.yml [hunk 9]
```diff
@@ -26,7 +26,7 @@ jobs:
     needs: call-create-release-workflow
     if: ${{ needs.call-create-release-workflow.outputs.is-release }}
     # owner/repository of workflow has to be static, see https://github.community/t/env-variables-in-uses/17466
-    uses: iotaledger/identity.rs/.github/workflows/shared-build-wasm.yml@main
+    uses: iotaledger/identity.rs/.github/workflows/shared-build-wasm.yml@typedoc
```
Comment (Dr-Electron) [inline .github/workflows/wasm-automatic-release-and-publish.yml:29]:
```suggestion
    uses: iotaledger/identity.rs/.github/workflows/shared-build-wasm.yml@main
```

File: .github/workflows/wasm-publish-to-npm.yml [hunk 10]
```diff
@@ -19,7 +19,7 @@ jobs:
 
   build-wasm: 
     # owner/repository of workflow has to be static, see https://github.community/t/env-variables-in-uses/17466
-    uses: iotaledger/identity.rs/.github/workflows/shared-build-wasm.yml@main
+    uses: iotaledger/identity.rs/.github/workflows/shared-build-wasm.yml@typedoc
```
Comment (Dr-Electron) [inline .github/workflows/wasm-publish-to-npm.yml:22]:
```suggestion
    uses: iotaledger/identity.rs/.github/workflows/shared-build-wasm.yml@main
```

**Review by Dr-Electron — APPROVED** [review]

**Review by UMR1352 — APPROVED** [review]
